pilot/pkg/xds/discovery.go

```diff
@@ -528,11 +523,9 @@ func (s *DiscoveryServer) initPushContext(req *model.PushRequest, oldPushContext
 		return nil, err
 	}
 
-	s.updateMutex.Lock()
-	s.Env.PushContext = push
 	// Ensure we drop the cache in the lock to avoid races, where we drop the cache, fill it back up, then update push context
```

Comment:
This is broken, as noted it must be done with the lock held

Reply:
I think not, here we update pushContext first then clear cache

Reply:
Then a generator generate xds with old pc would not be storred. It can prevent using stale cache

Reply:
At the very least the comment is wrong since it says we drop the caching _in the lock_.

But I think its a real issue.

Between L527 and L528 we can have the cache populate with the old PC. It needs to be atomic

Reply:
Let me think about it carefully

Reply:
>Between L527 and L528 we can have the cache populate with the old PC. It needs to be atomic

We have token in the cache, if we drop cache in L527, we updated the token to the current time. And a old PC generated xds can not be storred because the old token is staled

Reply:
I see. I think it can still be violated from this:

```
func AdsPushAll(s *DiscoveryServer) {
```

But its only used for debug so not a huge deal